src/mame/tektronix/tek440x.cpp

```diff
@@ -57,17 +58,295 @@
+		if (BIT(*m_map_control, MAP_VM_ENABLE))
+		{
+			if (intention == TR_WRITE)
+			if (BIT(m_map[address >> 12], 14) == 0)	// read only
+			{
+				LOG("memory_translate: write fail\n");
+				return false;
+			}
+
+			// dont try and translate a null page
+			if (BIT(m_map[address >> 12], 11, 3) != 0)
+			{
+				LOG("memory_translate: map %08x => paddr(%08x)\n",(address), (BIT(address, 0, 12) | (BIT(m_map[address >> 12], 0, 11) << 12) ) );
+			
+				address = BIT(address, 0, 12) | (BIT(m_map[address >> 12], 0, 11) << 12);
+			}
+		}
+		
+		return true;
+	}
+
+	u32 mmu_translate_address(offs_t address, u8 rw)
+	{
+	
+		if (!buserror_occurred)
+		if (*m_map_control & (1 << MAP_BLOCK_ACCESS))
+		if ((get_fc() & 4) == 0)				// only in User mode
+		if (BIT(*m_map_control, MAP_VM_ENABLE))
+		{
+			LOG("mmu_translate_address: map %08x => paddr(%08x) fc(%d) pc(%08x)\n",(address), BIT(address, 0, 12) | (BIT(m_map[address >> 12], 0, 11) << 12), get_fc(), pc());
+
+			if (rw)
+			{
+				// is !cpuWr
+				*m_map_control &= ~(1 << MAP_CPU_WR);
+			}
+			else
+			{
+				// is cpuWr
+				*m_map_control |= (1 << MAP_CPU_WR);
+			}
+
+			// matching pid
+			if (BIT(m_map[address >> 12], 11, 3) != (*m_map_control & 7))
+			{
+				*m_map_control &= ~(1 << MAP_BLOCK_ACCESS);
+
+				LOG("mmu_translate_address: bus error: PID(%d) != %d %08x fc(%d) pc(%08x)\n", BIT(m_map[address >> 12], 11, 3), (*m_map_control & 7), address, get_fc(), pc());
+				buserror_occurred = true;
+				set_buserror_details(address, rw, get_fc(), true);
+				return address;
+			}
+			else
+			{
+				*m_map_control |= (1 << MAP_BLOCK_ACCESS);
+			}
+
+			// write enabled?
+			if ((rw==0) && BIT(m_map[address >> 12], 14) == 0)	// read only page
+			{
+				*m_map_control &= ~(1 << MAP_BLOCK_ACCESS);
+
+				LOG("mmu_translate_address: bus error: %08x fc(%d) pc(%08x)\n",(address), get_fc(), pc());
+				buserror_occurred = true;
+				set_buserror_details(address, rw, get_fc(), true);
+				return address;
+			}
+		
+			address = BIT(address, 0, 12) | (BIT(m_map[address >> 12], 0, 11) << 12);
+		}
+
+		buserror_occurred = false;
+
+		return address;
+	}
+
+u16 PTEread(offs_t address)
+{
+	// selftest does a read and expects it to fail iff !MAP_SYS_WR_ENABLE; its not WR enable, its enable..
+	if (!BIT(*m_map_control, MAP_SYS_WR_ENABLE))
+	{
+			LOG("PTEread: bus error: PID(%d) %08x fc(%d) pc(%08x)\n", BIT(m_map[(address >> 12) & 0x7ff], 11, 3), (address), get_fc(), pc());
+			buserror_occurred = true;
+			set_buserror_details(address, 1, get_fc(), true);
+			return 0;
+	}
+
+	return m_map[(address>>12) & 0x7ff];
+}
+
+void PTEwrite(offs_t address, u16 data)
+{
+	if (((address>>12) & 0x7ff) < 20)
+		LOG("PTEwrite: %08x  <= %04x paddr(%08x) PID(%d) dirty(%d) write_enable(%d)\n",
+		(address>>12) & 0x7ff, data,
+		(BIT(data, 0, 12)<<12), BIT(data, 11, 3), data & 0x8000 ? 1 : 0, data & 0x4000 ? 1 : 0);
+
+	if (BIT(*m_map_control, MAP_SYS_WR_ENABLE))
+	{
+		m_map[(address>>12) & 0x7ff] = data;
+	}
+}
```

Comment:
Indentation is off here:
* These functions are at the left margin while the other inline member function bodies are indented.
* You should use one level of indentation per nested scope level.
* Please use braces for any multi-line `if`/`else` body even if it’s technically only a single statement.
* Please indent statement continuations by two tabs.